crates/router/src/connector/bluesnap.rs

```diff
@@ -0,0 +1,741 @@
+
+use std::fmt::Debug;
+
+use base64::Engine;
+use common_utils::crypto;
+use error_stack::{IntoReport, ResultExt};
+use transformers as bluesnap;
+
+use crate::{
+    configs::settings,
+    consts,
+    core::{
+        errors::{self, CustomResult},
+        payments,
+    },
+    db::StorageInterface,
+    headers, logger,
+    services::{self, ConnectorIntegration},
+    types::{
+        self,
+        api::{self, ConnectorCommon, ConnectorCommonExt},
+        ErrorResponse, Response,
+    },
+    utils::{self, BytesExt},
+};
+
+#[derive(Debug, Clone)]
+pub struct Bluesnap;
+
+impl<Flow, Request, Response> ConnectorCommonExt<Flow, Request, Response> for Bluesnap
+where
+    Self: ConnectorIntegration<Flow, Request, Response>,
+{
+    fn build_headers(
+        &self,
+        req: &types::RouterData<Flow, Request, Response>,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+        let mut header = self.get_auth_header(&req.connector_auth_type)?;
+        header.push((
+            headers::CONTENT_TYPE.to_string(),
+            self.get_content_type().to_string(),
+        ));
+        Ok(header)
+    }
+}
+
+impl ConnectorCommon for Bluesnap {
+    fn id(&self) -> &'static str {
+        "bluesnap"
+    }
+
+    fn common_get_content_type(&self) -> &'static str {
+        "application/json"
+    }
+
+    fn base_url<'a>(&self, connectors: &'a settings::Connectors) -> &'a str {
+        connectors.bluesnap.base_url.as_ref()
+    }
+
+    fn get_auth_header(
+        &self,
+        auth_type: &types::ConnectorAuthType,
+    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+        let auth: bluesnap::BluesnapAuthType = auth_type
+            .try_into()
+            .change_context(errors::ConnectorError::FailedToObtainAuthType)?;
+        let encoded_api_key =
+            consts::BASE64_ENGINE.encode(format!("{}:{}", auth.key1, auth.api_key));
+        Ok(vec![(
+            headers::AUTHORIZATION.to_string(),
+            format!("Basic {}", encoded_api_key),
+        )])
+    }
+
+    fn build_error_response(
+        &self,
+        res: Response,
+    ) -> CustomResult<ErrorResponse, errors::ConnectorError> {
+        logger::debug!(bluesnap_error_response=?res);
+        let response: bluesnap::BluesnapErrorResponse = res
+            .response
+            .parse_struct("BluesnapErrorResponse")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+        Ok(ErrorResponse {
+            status_code: res.status_code,
+            code: match response.message.first() {
+                Some(first_error) => first_error.code.clone(),
+                _ => consts::NO_ERROR_CODE.to_string(),
+            },
+            message: match response.message.first() {
+                Some(first_error) => first_error.description.clone(),
+                _ => consts::NO_ERROR_MESSAGE.to_string(),
+            },
```

Comment:
repeated code, use a top level match and populate fields only if error is present